src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala

```diff
@@ -0,0 +1,198 @@
+import org.tensorframes.impl.{DebugRowOps, SerializedGraph, TensorFlowOps}
+
+
+class DeepImageFeaturizer(override val uid: String) extends Transformer with DefaultParamsWritable {
+
+  def this() = this(Identifiable.randomUID("deepImageFeaturizer"))
+
+  final val inputCol: Param[String] = new Param[String](this, "inputCol", "input column name")
+  final val outputCol: Param[String] = new Param[String](this, "outputCol", "output column name")
+
+  // TODO: make modelName a param
+  val modelName = "_test"
+  def getModelName: String = modelName
+
+  private val RESIZED_IMAGE_COL = "__sparkdl_imageResized"
+  private val INPUT_BUFFER_COL = "__sparkdl_imageBuffer"
+
+  private def validateSchema(schema: StructType) {
+    val inputColumnName = getInputCol
+    require(
+      schema.fieldNames.contains(inputColumnName),
+      s"Input DataFrame must contain column named $inputColumnName")
+    require(
+      ! schema.fieldNames.contains(getOutputCol),
+      s"Input DataFrame cannot already contain a column with name $getOutputCol"
+    )
+    val fieldIndex = schema.fieldIndex(inputColumnName)
+    val colType = schema.fields(fieldIndex).dataType
+    require(
+      colType == ImageSchema.columnSchema,
+      s"inputCol must be an image column with schema ImageSchema.columnSchema, got ${colType}"
+    )
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    validateSchema(schema)
+    schema.add(getOutputCol, VectorType)
+  }
+
+  override def copy(extra: ParamMap): Transformer = {
+    defaultCopy(extra)
+  }
+
+  def getInputCol: String = {
+    getOrDefault(inputCol)
+  }
+
+  def getOutputCol: String = {
+    getOrDefault(outputCol)
+  }
+
+  def setInputCol(value: String): this.type = {
+    set(inputCol, value)
+    this
+  }
+
+  def setOutputCol(value: String): this.type = {
+    set(outputCol, value)
+    this
+  }
+
+  def transform(dataFrame: Dataset[_]): DataFrame = {
+    validateSchema(dataFrame.schema)
+    val model = DeepImageFeaturizer._supportedModels.getOrElse(
+      getModelName,
+      throw new IllegalArgumentException(s"Unknown model: $getModelName")
+    )
+
+    val imSchema = ImageSchema.columnSchema
+    val height = model.height
+    val width = model.width
+    val resizeUdf = udf((image: Row) => ImageUtils.resizeImage(height, width, 3, image), imSchema)
+
+    val imageDF = dataFrame
+      .withColumn(RESIZED_IMAGE_COL, resizeUdf(col(getInputCol)))
+      .withColumn(INPUT_BUFFER_COL, col(s"$RESIZED_IMAGE_COL.data"))
+
+    val shapeHints = new ShapeDescription(
+      /*
+      // TODO: map doesn't seem to matter
+      out = Map("input" -> Shape(299 * 299 * 3), "output" -> Shape(1234)),
+       */
+      out = Map.empty[String, Shape],
+      requestedFetches = Seq(model.graphOutputNode),
+      inputs = Map(model.graphInputNode -> INPUT_BUFFER_COL)
+    )
+
+    val toVector = udf { features: Seq[Double] =>
+      Vectors.dense(features.toArray)
+    }
+
+    DebugRowOps
```

Comment:
why is this called `Debug...`? xD @thunterdb

Reply:
Yes @thunterdb, why is it named Debug... :).

Reply:
Well, it is called debug, because it was written with idea that it would be soon replaced by `OptimizedRowOps`, but got optimized enough in the process...

You are not supposed to access this internal function, but at the same time it seems useful to expose it.